torch/quantization/fx/_equalize.py

```diff
@@ -224,3 +212,259 @@ def node_supports_equalization(node: Node, modules) -> bool:
+    # If the following layer needs to be equalized then we will multiply its scale
+    weight = modules[node.target].weight
+    assert(isinstance(weight, torch.Tensor))
+
+    scaled_weight = torch.mul(weight, torch.reciprocal(equalization_scale))
+
+    if next_equalization_scale is None:
+        modules[node.target].weight = nn.Parameter(scaled_weight)
+        return
+
+    scaled_weight = torch.mul(scaled_weight, next_equalization_scale)
+    modules[node.target].weight = nn.Parameter(scaled_weight)
+
+    # TODO: The bias may need to be scaled for connecting linear layers
+    bias = modules[node.target].bias
+    assert(isinstance(bias, torch.Tensor))
+
+    scaled_bias = torch.mul(bias, next_equalization_scale)
+    modules[node.target].bias = nn.Parameter(scaled_bias)
+
+def update_obs_for_equalization(model: GraphModule, modules: Dict[str, nn.Module]) -> Dict[str, _WeightEqualizationObserver]:
+    """ Update all of the observer's equalization scale. For each
+    InputEqualizationObserver, we will find the location of the next
+    WeightEqualizationObserver, create it, and calculate the equalization scale
+    based on the two observers.
+
+    We will then return a dictionary mapping node names to the newly created
+    WeightEqualizationObserver.
+    """
+    weight_eq_obs_dict = {}
+    for node in model.graph.nodes:
+        if node.op == 'call_module' and isinstance(modules[node.target], _InputEqualizationObserver):
+            input_eq_obs = modules[node.target]
+            assert(isinstance(input_eq_obs, _InputEqualizationObserver))
+            next_node, weight_eq_obs = get_weight_eq_obs(node, model, modules)
+
+            if next_node is None or weight_eq_obs is None:
+                continue
+
+            weight_eq_obs(modules[str(next_node.target)].weight)
+
+            # Calculate and set the equalization scale values
+            equalization_scale = calculate_equalization_scale(input_eq_obs, weight_eq_obs)
+            input_eq_obs.set_equalization_scale(equalization_scale)
+            weight_eq_obs.set_equalization_scale(equalization_scale)
+
+            weight_eq_obs_dict[next_node.name] = weight_eq_obs
+
+    return weight_eq_obs_dict
+
+def convert_eq_obs(
+    model: GraphModule,
+    modules: Dict[str, nn.Module],
+    weight_eq_obs_dict: Dict[str, _WeightEqualizationObserver],
+) -> None:
+    """ Removes the input equalization observers and replaces them with mul
+    operators whenever applicable. Updates the input quantization observers with
+    the scaled input min/max values. Scales the weights by the current and next
+    equalization scales, and removes the weight equalization observer node if it
+    exists.
+
+    Before:
+                                    weight values
+                                          |
+                                    WeightQuantObs
+                                          |
+                                      WeightEqObs
```

Comment:
Yes, from the discussion here: https://github.com/pytorch/pytorch/pull/59747#discussion_r651976455. Right now the equalization nodes do a "swapping" during convert:
```
# After prepare
x0 -> quant_obs -> eq_obs -> linear1

# After convert
x0 -> mul -> quantize_per_tensor -> quantized::linear1
```
I think some of the issue with having the `eq_obs` before the `quant_obs` is that it would modify the graph before the current node inserts its observer, and cause some inconsistencies the ordering of the observers in relation to each other (https://github.com/pytorch/pytorch/pull/59963#discussion_r651357501).

Reply:
thanks for explaining, @angelayi .  Can we add this explanation to the comments of the code, since most code readers will probably have this question?